File: cordova-plugin-prayer-widget/src/android/PrayerWidget.java
package com.dkurve.betterislamqa.prayerwidget;

import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CallbackContext;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import android.content.Context;
import android.content.SharedPreferences;
import android.appwidget.AppWidgetManager;
import android.content.ComponentName;
import android.content.Intent;
import android.util.Log;

/**
 * PrayerWidget Cordova Plugin
 * Bridges JavaScript calls to native Android widget updates
 */
public class PrayerWidget extends CordovaPlugin {

    private static final String TAG = "PrayerWidget";
    private static final String PREFS_NAME = "PrayerWidgetPrefs";

    @Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext)
            throws JSONException {

        Log.d(TAG, "Execute action: " + action);

        if (action.equals("updateWidget")) {
            JSONObject prayerData = args.getJSONObject(0);
            this.updateWidget(prayerData, callbackContext);
            return true;
        }
        else if (action.equals("isWidgetInstalled")) {
            this.isWidgetInstalled(callbackContext);
            return true;
        }
        else if (action.equals("forceUpdate")) {
            this.forceUpdate(callbackContext);
            return true;
        }
        else if (action.equals("clearWidget")) {
            this.clearWidget(callbackContext);
            return true;
        }
        else if (action.equals("consumeLaunchAction")) {
            this.consumeLaunchAction(callbackContext);
            return true;
        }

        return false;
    }

    /**
     * Update widget with new prayer time data
     */
    private void updateWidget(JSONObject prayerData, CallbackContext callbackContext) {
        Context context = cordova.getActivity().getApplicationContext();

        // Save data to SharedPreferences so widget can access it
        SharedPreferences prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = prefs.edit();

        // All prayer times
        editor.putString("fajr_time", prayerData.optString("fajr", "--:--"));
        editor.putString("dhuhr_time", prayerData.optString("dhuhr", "--:--"));
        editor.putString("asr_time", prayerData.optString("asr", "--:--"));
        editor.putString("maghrib_time", prayerData.optString("maghrib", "--:--"));
        editor.putString("isha_time", prayerData.optString("isha", "--:--"));

        // Current/Next prayer info
        editor.putString("next_prayer", prayerData.optString("nextPrayer", "Fajr"));
        editor.putString("next_prayer_time", prayerData.optString("nextPrayerTime", "--:--"));
        editor.putString("time_remaining", prayerData.optString("timeRemaining", "--:--"));
        editor.putString("current_prayer", prayerData.optString("currentPrayer", ""));
        editor.putString("current_prayer_end", prayerData.optString("currentPrayerEnd", ""));
        editor.putInt("schema_version", prayerData.optInt("schemaVersion", 1));
        editor.putString("timezone", prayerData.optString("timezone", ""));
        editor.putString("location_name", prayerData.optString("locationName", ""));
        editor.putLong("generated_at_ms", optEpochMillis(prayerData, "generatedAtMs"));
        JSONObject timestamps = prayerData.optJSONObject("prayerTimestamps");
        if (timestamps == null) timestamps = prayerData;
        editor.putLong("fajr_timestamp", optEpochMillis(timestamps, "fajr"));
        editor.putLong("sunrise_timestamp", optEpochMillis(timestamps, "sunrise"));
        editor.putLong("dhuhr_timestamp", optEpochMillis(timestamps, "dhuhr"));
        editor.putLong("asr_timestamp", optEpochMillis(timestamps, "asr"));
        editor.putLong("maghrib_timestamp", optEpochMillis(timestamps, "maghrib"));
        editor.putLong("isha_timestamp", optEpochMillis(timestamps, "isha"));
        editor.putLong("next_fajr_timestamp", optEpochMillis(timestamps, "nextFajr"));
        editor.putLong("last_updated", System.currentTimeMillis());
        editor.apply();

        Log.d(TAG, "Widget data saved: " + prayerData.toString());

        // Trigger widget update
        this.triggerWidgetUpdate(context);

        callbackContext.success("Widget updated successfully");
    }

    private long optEpochMillis(JSONObject data, String key) {
        long value = data.optLong(key, 0L);
        // Reject seconds, negative values, and implausibly distant instants.
        long now = System.currentTimeMillis();
        long tenYears = 10L * 365L * 24L * 60L * 60L * 1000L;
        return value > 1000000000000L && Math.abs(value - now) < tenYears ? value : 0L;
    }

    /**
     * Return and clear the action attached to the Activity launch Intent.
     * The web layer should call this on deviceready and resume, then route the
     * value "prayer-times" to its prayer-times screen.
     */
    private void consumeLaunchAction(CallbackContext callbackContext) {
        Intent launchIntent = cordova.getActivity().getIntent();
        boolean openPrayerTimes = launchIntent != null
            && launchIntent.getBooleanExtra("openPrayerTimes", false);

        if (openPrayerTimes) {
            launchIntent.removeExtra("openPrayerTimes");
            callbackContext.success("prayer-times");
        } else {
            callbackContext.success("");
        }
    }

    /**
     * Check if widget is installed (added to home screen)
     */
    private void isWidgetInstalled(CallbackContext callbackContext) {
        try {
            Context context = cordova.getActivity().getApplicationContext();
            AppWidgetManager appWidgetManager = AppWidgetManager.getInstance(context);

            int[] ids = appWidgetManager.getAppWidgetIds(
                new ComponentName(context, PrayerTimeWidgetProvider.class));

            boolean isInstalled = ids.length > 0;
            Log.d(TAG, "Widget installed: " + isInstalled + " (count: " + ids.length + ")");

            callbackContext.success(isInstalled ? 1 : 0);
        } catch (Exception e) {
            Log.e(TAG, "Error checking widget installation: " + e.getMessage());
            callbackContext.error("Failed to check widget installation: " + e.getMessage());
        }
    }

    /**
     * Force widget update (useful for testing)
     */
    private void forceUpdate(CallbackContext callbackContext) {
        try {
            Context context = cordova.getActivity().getApplicationContext();
            this.triggerWidgetUpdate(context);
            callbackContext.success("Widget force update triggered");
        } catch (Exception e) {
            Log.e(TAG, "Error forcing widget update: " + e.getMessage());
            callbackContext.error("Failed to force widget update: " + e.getMessage());
        }
    }

    private void clearWidget(CallbackContext callbackContext) {
        try {
            Context context = cordova.getActivity().getApplicationContext();
            context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE).edit().clear().apply();
            this.triggerWidgetUpdate(context);
            callbackContext.success("Widget data cleared");
        } catch (Exception e) {
            callbackContext.error("Failed to clear widget data: " + e.getMessage());
        }
    }

    /**
     * Trigger widget update broadcast
     */
    private void triggerWidgetUpdate(Context context) {
        Intent intent = new Intent(context, PrayerTimeWidgetProvider.class);
        intent.setAction(AppWidgetManager.ACTION_APPWIDGET_UPDATE);

        AppWidgetManager appWidgetManager = AppWidgetManager.getInstance(context);
        int[] ids = appWidgetManager.getAppWidgetIds(
            new ComponentName(context, PrayerTimeWidgetProvider.class));

        intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_IDS, ids);
        context.sendBroadcast(intent);

        Log.d(TAG, "Widget update broadcast sent to " + ids.length + " widgets");
    }
}

File: cordova-plugin-prayer-widget/src/ios/PrayerWidget.swift
import Foundation
import WidgetKit

/**
 * PrayerWidget Cordova Plugin - iOS Implementation
 * Bridges JavaScript calls to native iOS widget updates
 */
@objc(PrayerWidget)
class PrayerWidget: CDVPlugin {

    private let appGroupName = "group.com.dkurve.betterislamqa"

    /**
     * Update widget with new prayer time data
     */
    @objc(updateWidget:)
    func updateWidget(command: CDVInvokedUrlCommand) {
        var pluginResult = CDVPluginResult(status: CDVCommandStatus_ERROR)

        guard let prayerData = command.arguments[0] as? [String: Any] else {
            pluginResult = CDVPluginResult(
                status: CDVCommandStatus_ERROR,
                messageAs: "Invalid prayer data"
            )
            self.commandDelegate!.send(pluginResult, callbackId: command.callbackId)
            return
        }

        // Save data to App Group shared container
        if let userDefaults = UserDefaults(suiteName: appGroupName) {
            // All 5 prayer times
            userDefaults.set(prayerData["fajr"] as? String ?? "--:--", forKey: "fajr_time")
            userDefaults.set(prayerData["dhuhr"] as? String ?? "--:--", forKey: "dhuhr_time")
            userDefaults.set(prayerData["asr"] as? String ?? "--:--", forKey: "asr_time")
            userDefaults.set(prayerData["maghrib"] as? String ?? "--:--", forKey: "maghrib_time")
            userDefaults.set(prayerData["isha"] as? String ?? "--:--", forKey: "isha_time")

            // Current/Next prayer info
            userDefaults.set(prayerData["nextPrayer"] as? String ?? "Fajr", forKey: "next_prayer")
            userDefaults.set(prayerData["nextPrayerTime"] as? String ?? "--:--", forKey: "next_prayer_time")
            userDefaults.set(prayerData["timeRemaining"] as? String ?? "--:--", forKey: "time_remaining")
            userDefaults.set(prayerData["currentPrayer"] as? String ?? "", forKey: "current_prayer")
            userDefaults.set(prayerData["currentPrayerEnd"] as? String ?? "", forKey: "current_prayer_end")
            userDefaults.set(prayerData["schemaVersion"] as? Int ?? 1, forKey: "schema_version")
            userDefaults.set(prayerData["timezone"] as? String ?? "", forKey: "timezone")
            userDefaults.set(prayerData["locationName"] as? String ?? "", forKey: "location_name")
            persistEpoch(prayerData, key: "generatedAtMs", defaultsKey: "generated_at_ms", defaults: userDefaults)
            let timestamps = prayerData["prayerTimestamps"] as? [String: Any] ?? prayerData
            persistEpoch(timestamps, key: "fajr", defaultsKey: "fajr_timestamp", defaults: userDefaults)
            persistEpoch(timestamps, key: "sunrise", defaultsKey: "sunrise_timestamp", defaults: userDefaults)
            persistEpoch(timestamps, key: "dhuhr", defaultsKey: "dhuhr_timestamp", defaults: userDefaults)
            persistEpoch(timestamps, key: "asr", defaultsKey: "asr_timestamp", defaults: userDefaults)
            persistEpoch(timestamps, key: "maghrib", defaultsKey: "maghrib_timestamp", defaults: userDefaults)
            persistEpoch(timestamps, key: "isha", defaultsKey: "isha_timestamp", defaults: userDefaults)
            persistEpoch(timestamps, key: "nextFajr", defaultsKey: "next_fajr_timestamp", defaults: userDefaults)
            userDefaults.set(Date().timeIntervalSince1970, forKey: "last_updated")
            userDefaults.synchronize()

            // Reload all widgets
            if #available(iOS 14.0, *) {
                WidgetCenter.shared.reloadAllTimelines()
                NSLog("PrayerWidget: Widget data updated and timelines reloaded")
            }

            pluginResult = CDVPluginResult(
                status: CDVCommandStatus_OK,
                messageAs: "Widget updated successfully"
            )
        } else {
            pluginResult = CDVPluginResult(
                status: CDVCommandStatus_ERROR,
                messageAs: "Failed to access shared container"
            )
        }

        self.commandDelegate!.send(pluginResult, callbackId: command.callbackId)
    }

    private func persistEpoch(_ data: [String: Any], key: String, defaultsKey: String, defaults: UserDefaults) {
        let milliseconds = (data[key] as? NSNumber)?.doubleValue ?? 0
        let now = Date().timeIntervalSince1970 * 1000
        let tenYears = 10.0 * 365.0 * 24.0 * 60.0 * 60.0 * 1000.0
        defaults.set(milliseconds > 1_000_000_000_000 && abs(milliseconds - now) < tenYears ? milliseconds : 0,
                     forKey: defaultsKey)
    }

    /**
     * Check if widget is installed (always returns true on iOS as there's no way to detect)
     */
    @objc(isWidgetInstalled:)
    func isWidgetInstalled(command: CDVInvokedUrlCommand) {
        // iOS doesn't provide an API to check if a widget is added to home screen
        // We assume widgets might be installed if the app is installed
        let pluginResult = CDVPluginResult(
            status: CDVCommandStatus_OK,
            messageAs: 1
        )
        self.commandDelegate!.send(pluginResult, callbackId: command.callbackId)
    }

    /**
     * Force widget update
     */
    @objc(forceUpdate:)
    func forceUpdate(command: CDVInvokedUrlCommand) {
        if #available(iOS 14.0, *) {
            WidgetCenter.shared.reloadAllTimelines()
            NSLog("PrayerWidget: Force update triggered")

            let pluginResult = CDVPluginResult(
                status: CDVCommandStatus_OK,
                messageAs: "Widget force update triggered"
            )
            self.commandDelegate!.send(pluginResult, callbackId: command.callbackId)
        } else {
            let pluginResult = CDVPluginResult(
                status: CDVCommandStatus_ERROR,
                messageAs: "WidgetKit not available (iOS 14+ required)"
            )
            self.commandDelegate!.send(pluginResult, callbackId: command.callbackId)
        }
    }

    @objc(clearWidget:)
    func clearWidget(command: CDVInvokedUrlCommand) {
        guard let defaults = UserDefaults(suiteName: appGroupName) else {
            self.commandDelegate!.send(
                CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: "Failed to access shared container"),
                callbackId: command.callbackId
            )
            return
        }
        for key in defaults.dictionaryRepresentation().keys {
            defaults.removeObject(forKey: key)
        }
        if #available(iOS 14.0, *) {
            WidgetCenter.shared.reloadAllTimelines()
        }
        self.commandDelegate!.send(
            CDVPluginResult(status: CDVCommandStatus_OK, messageAs: "Widget data cleared"),
            callbackId: command.callbackId
        )
    }

    /** Android widget launches can carry an explicit route. The iOS extension
     * currently opens the host app without a route, so expose the same API with
     * an empty one-shot action rather than failing the cross-platform call. */
    @objc(consumeLaunchAction:)
    func consumeLaunchAction(command: CDVInvokedUrlCommand) {
        let pluginResult = CDVPluginResult(
            status: CDVCommandStatus_OK,
            messageAs: ""
        )
        self.commandDelegate!.send(pluginResult, callbackId: command.callbackId)
    }
}
